### ISSUE TITLE
Discussion: Storing Big Data

### ISSUE THREAD

**Issue by thomasaarholt** [hyperspy/hyperspy#1804]
Recently I've been thinking a bit about the future of big data in Electron Microscopy. 

For the uninitiated: Due to increases in detector resolution and the increase of use of time-resolved EM, 4D STEM and simultaneous acquisition of multi-frame imaging and spectrum-imaging (and more), datasets are becoming _enormous_, and will soon be approaching __terabytes__ per session (if they aren't already).

As an example, a 1024 * 1024 pixel EELS spectrum image with 4096 channels and a double precision (8 bytes per "pixel") takes up `1024 * 1024 * 4096 * 8 = 34.4 GB`. I was acquiring this type of dataset towards the end of my PhD. 

We are quickly becoming our local IT person's worst nightmare, and it is our responsibility and to our advantage to handle this well.

Hyperspy already has implemented _lazy loading_, a method of only loading into memory requested data, rather than loading a whole dataset, a technique essential to analysing such vast amounts of data. However, I'd like to focus this discussion on the **storage** of such data.

- What measures are labs already taking to handle large amounts of data, and how are they working?
- What features does hyperspy already have for size reduction? Compression? Dimensionality reduction?
- What new algorithms for handling large data could be of interest?
- How can we improve hyperspy's current functionality?

**Comment by thomasaarholt** [comment]
My own thoughts centre around compression and dimensionality reduction.
I know that [hdf5 supports compression](https://support.hdfgroup.org/HDF5/faq/compression.html), but our own documentation doesn't refer to it.
- Does hyperspy use compression with hdf5?

When performing PCA or other dimensionality reduction algorithms, the result is a list of pairs of factors and loadings. When saved (`s.export_decomposition_results()`), these appear as individual files in a folder, and are not user friendly at all. 

- Can we come up with a better way to save the learning results in a single dataset that can be stored, and then loaded (regular or lazily) into the reconstructed dataset?

**Comment by robbmcleod** [comment]
The cryo-EM community routinely deals with TB per session.  I can try and relay some of the lessions I've learned. I wrote a short comm for J. Struct. Bio. on using high-speed compressors with sparse data:

https://www.sciencedirect.com/science/article/pii/S1047847717302083

Source:

https://github.com/em-MRCZ

Here's the PR for adding a plugin for MRCZ:

https://github.com/hyperspy/hyperspy/pull/1716

Efforts to build a plugin for Hyperspy are a little stranded by the difficulty we are having getting TravisCI to build `blosc`, the compression library, since it uses AVX2 instructions.  `blosc` is the de-facto standard plugin for HDF5 compression as well, so getting it built with `manywheels` support would be generally helpful.  Here's our efforts to get blosc building with TravisCI so wheels can be uploaded to PyPI:

https://github.com/Blosc/python-blosc-wheels

https://github.com/robbmcleod/python-blosc-wheels

In the medium-term I expect when `c-blosc2` comes out to switch MRCZ onto that standard, at which point we can have sliced-access into the compressed array.

**Comment by magnunor** [comment]
We've been using in-built HDF5 compression to reduce the dataset sizes, which at least helps a bit: 8 GB raw binary data is reduced to ~2.3 GB.

**Comment by hentr** [comment]
I guess us here at Empa are not the only ones discussing putting EM data in openly accessible databases with conveniently available (and searchable) metadata. This is both for hosting the raw data used in articles, but it also enables big data studies of large volumes of data originating from multiple microscopes/institutions. This is important especially considering it is usually hard to do anything quantitative with EM that is macroscopically representative, since we're often too busy staring at areas in the nm scale. Open data is now also being pushed by many funding agencies, requiring an open data policy in grant applications. 

There are multiple good repositories for cryo-EM and 3D FIB-SEM, but these are all focused on life science data and seems to not be very suitable for more general EM data. Given the great ability of the hyperspy.io module at handling almost any EM data, is there anyone thinking of building an open database for EM data using the io module?

**Comment by magnunor** [comment]
From a pure software perspective this isn't really that difficult. I could probably spin up a system in a couple of hours using cloud-hosted GitLab and git-lfs, coupled with a bunch of prewritten docker templates to cover pretty much every processing regime. I've actually been thinking about something like this, but I don't really have the time at the moment.

The major issue is having the necessary storage space, as @thomasaarholt mentioned, especially the newer raw pixelated STEM data can easily be terabyte size. Zenodo offers free hosting of scientific data, which is where I'll upload my open data sets, but even they have a limit of 50 GB per dataset.

**Comment by thomasaarholt** [comment]
@hentr You're absolutely right to raise the topic! It might be a good idea to  list the currently existing infrastructure that's available for EM-related data. The two that I know of are:

- [eelsdb](https://eelsdb.eu/) (For single-spectra only, beautiful website but is in need of some maintenance)
- [EMPIAR / Electron Microscopy Public Image ARchive](https://www.ebi.ac.uk/pdbe/emdb/empiar/) (Aimed at stacks of 2D TEM images. Unsure about support for SIs.)

I feel like there's a serious grant needed to support such infrastructure. If we're talking permanent hosting of EM data, we'll be soon talking about 0.1-several TB per paper. That could easily push petabyte size in a year. Just up/downloading such data is a feat in itself. On a 100Mb/s connection a 2TB dataset takes nearly two days (meta: the black hole researchers who published yesterday shipped their crates of harddrives by ship from Antarctica because it was quicker).

Even more importantly, it needs to be done by an organisation willing to handle this data for a long time. The European BioInformatics Suite, which is the parent of EMPIAR (and is the cryo repo @hentr referred to, I think) manages that by advertising their strong big data / analysis / training capability, so that they have an activity that is centred around (and actively benefits from) such data storage.

@hentr Which is the repo for 3D FIB-SEM data?

**Comment by hentr** [comment]
I was indeed thinking of EMPIAR (which also has FIB-SEM data) and [EMDB](http://www.ebi.ac.uk/pdbe/emdb/index.html/)

**Comment by robbmcleod** [comment]
I'd recommend labs start talking to their institutions computer cluster management if they haven't already. From what I've seen academic clusters have lower prices for data storage than commercial solutions.

Also, compression with a modern codec like `zstandard` will make a big difference with sparse data like diffraction data, especially if you store the data as integers. I did recently get a PR into `pytables` to introduce AVX2 support for Windows, so it should be a bit faster now on that OS.
